test/e2e/kb/version_upgrade_test.go

```diff
@@ -0,0 +1,137 @@
+	"github.com/elastic/cloud-on-k8s/test/e2e/test/elasticsearch"
+	"github.com/elastic/cloud-on-k8s/test/e2e/test/kibana"
+	"github.com/stretchr/testify/assert"
+	v1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func TestVersionUpgrade(t *testing.T) {
+	name := "test-version-upgrade"
+	esBuilder := elasticsearch.NewBuilder(name).
+		WithESMasterDataNodes(1, elasticsearch.DefaultResources).
+		WithVersion("7.4.1")
+
+	kbBuilder := kibana.NewBuilder(name).
+		WithElasticsearchRef(esBuilder.Ref()).
+		WithNodeCount(3).
+		WithVersion("7.4.0")
+
+	opts := []client.ListOption{
+		client.InNamespace(kbBuilder.Kibana.Namespace),
+		client.MatchingLabels(map[string]string{
+			common.TypeLabelName:      label.Type,
+			label.KibanaNameLabelName: kbBuilder.Kibana.Name,
+		}),
+	}
+
+	test.RunMutationsWhileWatching(
+		t,
+		[]test.Builder{esBuilder, kbBuilder},
+		[]test.Builder{esBuilder, kbBuilder.WithVersion("7.4.1").WithMutatedFrom(&kbBuilder)},
+		[]test.Watcher{NewReadinessWatcher(opts...), NewVersionWatcher(opts...)},
+	)
+}
+
+func TestVersionUpgradeAndRespec(t *testing.T) {
+	name := "test-upgrade-and-respec"
+	esBuilder := elasticsearch.NewBuilder(name).
+		WithESMasterDataNodes(1, elasticsearch.DefaultResources).
+		WithVersion("7.4.1")
+
+	kbBuilder1 := kibana.NewBuilder(name).
+		WithElasticsearchRef(esBuilder.Ref()).
+		WithNodeCount(3).
+		WithVersion("7.4.0")
+
+	// Perform a Kibana version upgrade immediately followed by a Kibana configuration change.
+	// We want to make sure that the second upgrade will be done in rolling upgrade fashion instead of terminating
+	// and recreating all the pods at once.
+	kbBuilder2 := kbBuilder1.WithMutatedFrom(&kbBuilder1).WithVersion("7.4.1")
+	kbBuilder3 := kbBuilder2.WithMutatedFrom(&kbBuilder2).WithLabel("some", "label")
+
+	opts := []client.ListOption{
+		client.InNamespace(kbBuilder1.Kibana.Namespace),
+		client.MatchingLabels(map[string]string{
+			common.TypeLabelName:      label.Type,
+			label.KibanaNameLabelName: kbBuilder1.Kibana.Name,
+		}),
+	}
+
+	// checks whether after temporary downtime kibana will be available the rest of the time
+	var hadZero, shouldHaveNonZero, failed bool
+	w := test.NewWatcher("watch pods readiness", 1*time.Second,
+		func(k *test.K8sClient, t *testing.T) {
+			pods, err := k.GetPods(opts...)
+			if err != nil {
+				t.Logf("got error: %v", err)
+			}
+
+			ready := k8s.ReadyPodsCount(pods)
+			hadZero = hadZero || ready == 0
+			if hadZero && ready > 0 {
+				shouldHaveNonZero = true
+			}
+
+			if shouldHaveNonZero && ready == 0 {
+				failed = true
+			}
+		},
+		func(k *test.K8sClient, t *testing.T) {
+			assert.False(t, failed)
+		})
+
+	test.RunMutationsWhileWatching(
+		t,
+		[]test.Builder{esBuilder, kbBuilder1},
+		[]test.Builder{esBuilder, kbBuilder2, kbBuilder3},
+		[]test.Watcher{w, NewVersionWatcher(opts...)},
+	)
+}
+
+func NewReadinessWatcher(opts ...client.ListOption) test.Watcher {
```

Comment:
Maybe add a comment explaining the rationale, assumptions and invariants behind these two watchers